service/controller/app/v1/resource/configmap/current.go

```diff
@@ -34,25 +38,82 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 		return nil, nil
 	}
 
+	if cc.Status.TenantCluster.IsUnavailable {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster is unavailable")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		resourcecanceledcontext.SetCanceled(ctx)
+		return nil, nil
+	}
+
 	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding configmap %#q in namespace %#q", name, r.chartNamespace))
 
-	chart, err := cc.K8sClient.CoreV1().ConfigMaps(r.chartNamespace).Get(name, metav1.GetOptions{})
-	if apierrors.IsNotFound(err) {
+	ch := make(chan response)
+
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	go func() {
+		configmap, err := cc.K8sClient.CoreV1().ConfigMaps(r.chartNamespace).Get(name, metav1.GetOptions{})
+
+		ch <- response{
+			Configmap: configmap,
+			Error:     err,
+		}
+	}()
+
+	var res = response{}
+
+	select {
+	case res = <-ch:
+		// Fall through.
+	case <-ctx.Done():
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			// Set status so we don't try to connect to the tenant cluster
+			// again in this reconciliation loop.
+			cc.Status.TenantCluster.IsUnavailable = true
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "timeout getting configmap")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+			return nil, nil
+		}
+	}
+
+	if apierrors.IsNotFound(res.Error) {
 		// Return early as configmap does not exist.
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find configmap %#q in namespace %#q", name, r.chartNamespace))
 		return nil, nil
-	} else if tenant.IsAPINotAvailable(err) {
-		// We should not hammer tenant API if it is not available, the tenant cluster
-		// might be initializing. We will retry on next reconciliation loop.
+	} else if tenant.IsAPINotAvailable(res.Error) {
+		// Set status so we don't try to connect to the tenant cluster
+		// again in this reconciliation loop.
+		cc.Status.TenantCluster.IsUnavailable = true
```

Comment:
If we cancel the reconciliation, why we even need to set `cc.Status.TenantCluster.IsUnavailable` since it would eventually stop whole steps?

Reply:
Good point it should just cancel the resource not the whole loop. Fixed.

The following resources check the context and cancel the resource. To give us more control.